### PR TITLE
GH-112354: `END_FOR` instruction to only pop one value.

### PR DESCRIPTION
**Pull request by markshannon**
In order to support tier 2 side-exits from `FOR_ITER` variants, we need a target that has the same stack as the exhaustion guard.
Currently we can either jump to the `END_FOR`, as we do for generators or the instruction afterwards, which we do for `FOR_ITER`.
In order to exit a guard we need a target in between those instructions, that pops the iterator, but not the value.
The compiler emits END_FOR; POP_TOP instead of END_FOR. 
 Previously:
```
   END_FOR (Pops two values) <- FOR_ITEN_GEN targets this
   ...                       <- FOR_ITER targets this
```
Now:
``` 
   END_FOR (Pops one value)  <- FOR_ITEN_GEN targets this
   POP_TOP (Pops one value)  <- _GUARD_NOT_EXHAUSTED can target this
   ...                       <- FOR_ITER targets this
```

* Issue: gh-112354

### CODE REVIEW

**Comment by gvanrossum** [comment]
Maybe we can do without a separate END_FOR opcode and instead just generate a POP?

File: Include/internal/pycore_unicodeobject_generated.h [hunk 1]
```diff
@@ -669,6 +669,9 @@ _PyUnicode_InitStaticStrings(PyInterpreterState *interp) {
     string = &_Py_ID(aggregate_class);
     assert(_PyUnicode_CheckConsistency(string, 1));
     _PyUnicode_InternInPlace(interp, &string);
+    string = &_Py_ID(alias);
+    assert(_PyUnicode_CheckConsistency(string, 1));
+    _PyUnicode_InternInPlace(interp, &string);
```
Comment (gvanrossum) [inline Include/internal/pycore_unicodeobject_generated.h:674]: (The updates to this file (and a few other generated files) are spurious and will disappear after merging the latest main.)

**Review by gvanrossum — APPROVED** [review]
Looks good, never mind my premature commend about END_FOR/POP_TOP.

**Comment by markshannon** [comment]
[The benchmarks](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240117-3.13.0a3%2B-ed98265) show no performance impact.

**Comment by markshannon** [comment]
> EDIT: Never mind, it's needed for instrumentation.

That's evidently not clear. The order of the `POP_TOP` and `END_FOR` is important as well.
I'll add some comments.

**Review by gvanrossum — APPROVED** [review]
Thanks!

**Review by gvanrossum — APPROVED** [review]

**Comment by gvanrossum** [comment]
Oh, fun stuff. Windows compilation error. :-(

File: Misc/NEWS.d/next/Core and Builtins/2024-01-17-05-09-32.gh-issue-112354.Run9ko.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+The ``END_FOR`` instruction now pops only one value. This is to better
```
Comment (Eclips4) [inline Misc/NEWS.d/next/Core and Builtins/2024-01-17-05-09-32.gh-issue-112354.Run9ko.rst:1]: With this change, documentation for ``END_FOR`` should be updated.
Currently [it says](https://docs.python.org/3.13/library/dis.html#opcode-END_FOR): "Removes the top two values from the stack."

**Pull request merged**